### PR TITLE
NFC interface for library with samples

### PR DESCRIPTION
**Pull request by merlokk**
Added
* PCSCDevice for generic reader
* ACR1252UPCSCDevice for ACR1252U and compatible readers (mostly from [acs](https://www.acs.com.hk/en/products/342/acr1252u-usb-nfc-reader-iii-nfc-forum-certified-reader/))
* library almost not changed
* pyscard - optional dependency (in pcsc.py there is a flag UseNFC that can turn NFC off)

Examples:
* get_info_nfc - general NFC demo
* acr122u.py - work with acr122u reader and compatible via direct apdu channel
* acr1252u.py - work with acr1252u reader and compatible via control channel
* u2f_nfc.py - demo of u2f operations via NFC
* modified credential.py - work with fido2 via NFC

issue here: #47

### CODE REVIEW

**Comment by merlokk** [comment]
it seems all is ok.

**Comment by dainnilsson** [comment]
Thanks for your work on this. While I haven't yet had a chance to look at a deeper level into the code, I've had a cursory glance and have tested to see that it works on my machine (It does!). A few notes from my end:

- The dummy PCSCDevice implementation seems unnecessary, I'm perfectly happy with `import fido2.nfc` and `import fido2.pcsc` failing if pyscard isn't installed. As long as `CtapHidDevice` can still be used without pyscard that is sufficient. The examples can use something like:

```python
from fido2.hid import CtapHidDevice
dev = next(CtapHidDevice.list_devices(), None)
if dev is None:
    try:
        from fido2.nfc import CtapNfcDevice
        dev = next(CtapNfcDevice.list_devices(), None)
    except ImportError:
        pass  # No NFC support

if not dev:
    print('No device found')
...
```

- I'm reluctant to include features which are not strictly related to FIDO2, like LED control, and so on. I also don't want to have reader-specific classes. I'd prefer get rid of the abstract `_PCSCDevice` class and just have a single PCSCDevice implementation. Or, alternatively, to move the remaining code from `pcsc.py` into `nfc.py` and just use the `smartcard` module from the `CtapNfcDevice` class directly.

- Please change function names and variables to use `snake_case` rather than `CamelCase` for consistency with the rest of the library.

- `CtapNFCDevice` should be renamed to `CtapNfcDevice` for consistency with `CtapHidDevice`.

**Comment by merlokk** [comment]
camels and so on fixed.
readers stuff moved to another classes.

Why I think readers must be there: 
1. Some readers dont work with contactless cards by default. They need to transfer command via special commands or control sequences (like acr122u-sam or touchatag) and people needs to make them work. (ill add this ACR reader later). This readers are very common because of their's low price.
2. Some readers needs to be configured to on|off field and anti-collision card. sometimes they just dont do anything with card without configuration
3. Too complex code to make pos 1 and 2. needs to have too much experience to make some of readers to work
4. as a sample for developers. they will know how to add their specific reader. and know some "magic" constants. I see only 1 comment with this constant in the internet(

but... i can remove this code or move it to samples if you want

**Comment by dainnilsson** [comment]
Thank for the changes!

Getting the reader version, and the LED control code, is still not needed for FIDO functionality, correct? I would like to drop that from the library. I'd be ok with an example showing how to use this extra functionality, as `acr1252u.py` does, but keeping the reader specific code in that example file (invoking `PCSCDevice.control_exchange` manually, etc.), with comments describing what is done and why.

I would like to save the user from having to perform a bunch of magic ceremony for each reader, but the way I see people using this mainly is the developer wanting support as many readers as possible, and not necessarily knowing what type the end user has. Do you have any links to documentation about the specific configuration needed for these readers?

I'd prefer if the `CtapNfcDevices` returned by `list_devices` connected to the card in the constructor to avoid introducing the internal state. `CtapHidDevice` currently does this. This likely would mean adding some additional mechanism to list readers, as the user may need to be prompted to present the Authenticator if there is an NFC reader available. I think a reasonable approach would be that `PCSCDevice` connects to the card, and that `CtapNfcDevice` selects the applet (in their respective `__init__` methods), and no internal state variable would be needed.

**Comment by merlokk** [comment]
>Do you have any links to documentation about the specific configuration needed for these readers?
it from reader's SDK. like https://www.acs.com.hk/download-manual/1083/API-ACR122USAM-2.01.pdf
but sometimes it not have all needed information( then libs decompile or forums or etc..

**Comment by merlokk** [comment]
looks like all is ok

**Comment by syntaxcase** [comment]
Hi, cool stuff!

Here's some notes (that apply to the library code, not the examples):

- The logging module should be used instead of print, and in general I would reduce the number of logged messages.

- For consistency with the rest of the codebase ifs with multiple conditions should not have parentheses around each condition.

- There are multiple instances of `except Exception`; it would be better to catch the actual excpetion or at least the most specific one.

- When raising excpetions it would be better to create and raise specific ones, `Exception` is too generic.

- I did not have the time to test this yet, but my impression is that some code would work only with Python 3. The library must also work with Python 2, at least as long as it is supported.

File: fido2/nfc.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def __repr__(self):
+        return 'CtapNfcDevice(%s)' % self.descriptor
+
+    @property
+    def version(self):
+        """CTAP NFC protocol version.
+
+        :rtype: Tuple[int, int, int]
+        """
+        return [1, 0, 0]
```
Comment (syntaxcase) [inline fido2/nfc.py]: This should match the docstring and return a tuple.

Comment (merlokk) [reply fido2/nfc.py]: not a tuple and not a string(
current hid.py returns only one int

what to do? 
1. make as current, fix docstrings
2. make tuple[1,2,3] and fix hid.py?

i prefer 1st)

Comment (merlokk) [reply fido2/nfc.py]: Fixed 1st way

**Comment by merlokk** [comment]
`except Exception` - looks like `SmartcardException` is maximum. I can disassemble it, but there may be many exceptions and sometimes they not predictable.

About python2 - ill try it on holidays. some time ago it worked.

**Comment by merlokk** [comment]
looks like it cant import nfc module from python 2.7.

**Comment by merlokk** [comment]
now it works in python 2.7
but...
travis tests returns `AttributeError: 'module' object has no attribute 'test_nfc'`
local tests works OK
dont know....

**Comment by merlokk** [comment]
ok
i have checked - it works on python 2.7 and python3
so all is ok

File: fido2/nfc.py [hunk 2]
```diff
@@ -17,22 +17,23 @@ class CtapNfcDevice(CtapDevice):
     :cvar descriptor: Device descriptor.
     """
 
-    def __init__(self, descriptor, dev):
+    def __init__(self, descriptor, dev, no_card=False):
```
Comment (syntaxcase) [inline fido2/nfc.py]: Is `no_card` added only for the tests? I wouldn't change the API to ease testing, but provide a mocked `dev` with mocks for all the methods used in this constructor.

Comment (merlokk) [reply fido2/nfc.py]: ok, ill delete it

**Comment by merlokk** [comment]
done

**Comment by dainnilsson** [comment]
Sorry for this dragging on a big. I'm mostly happy with things now, though there are a few things I'll likely want to change. I think we need to do some more thinking around some parts though, so I'm leaning more towards merging this now into a development branch (see the `next` branch) and probably making some changes to it later (hope you don't mind!).

Would you mind rebasing off of the `next` branch and squashing your commits into a single one?

Also, as this is a substantial contribution, please state that you contribute this code under the license of the repository (2 clause BSD).

**Comment by merlokk** [comment]
I dont mind to merge it to `next` branch.  Maybe needs to update the next branch.

Its OK as for license. I know.

**Comment by merlokk** [comment]
@dainnilsson may you help me to squashing my commits into a single one?

**Comment by dainnilsson** [comment]
I can certainly do that, but I'm not sure if you'll still be listed as the author on the commit if I do, would you be ok with that?

**Comment by merlokk** [comment]
its ok

**Comment by dainnilsson** [comment]
Thanks for your contribution, these changes have now been merged into the `next` branch, which will contain active development until it is merged into `master`!

**Pull request closed without merging**